runtime/src/accounts_db.rs

```diff
@@ -3089,56 +3131,148 @@ impl AccountsDb {
+                    ScanStorageResult::Cached(_) => {
+                        panic!("Should not see cached keys in this `else` branch, since we checked this slot did not exist in the cache above");
+                    }
+                    ScanStorageResult::Stored(stored_keys) => {
+                        for set_ref in stored_keys.iter() {
+                            self.accounts_index.purge_exact(
+                                set_ref.key(),
+                                &purge_slot,
+                                &mut reclaims,
+                            );
+                        }
+                    }
+                }
+                purge_accounts_index_elapsed.stop();
+                purge_stats
+                    .purge_accounts_index_elapsed
+                    .fetch_add(purge_accounts_index_elapsed.as_us(), Ordering::Relaxed);
+
+                // `handle_reclaims()` should remove all the account index entries and
+                // storage entries
+                let mut handle_reclaims_elapsed = Measure::start("handle_reclaims_elapsed");
+                // Slot should be dead after removing all its account entries
+                let expected_dead_slot = Some(*remove_slot);
+                self.handle_reclaims(
+                    &reclaims,
+                    expected_dead_slot,
+                    Some(purge_stats),
+                    Some(&mut ReclaimResult::default()),
+                    false,
+                );
+                handle_reclaims_elapsed.stop();
+                purge_stats
+                    .handle_reclaims_elapsed
+                    .fetch_add(handle_reclaims_elapsed.as_us(), Ordering::Relaxed);
+                // After handling the reclaimed entries, this slot's
+                // storage entries should be purged from self.storage
+                assert!(self.storage.get_slot_stores(*remove_slot).is_none());
+            }
+
+            // It should not be possible that a slot is neither in the cache or storage. Even in
+            // a slot with all ticks, `Bank::new_from_parent()` immediately stores some sysvars
+            // on bank creation.
+        }
 
-                // Note this only cleans up the storage entries. The accounts index cleaning
-                // (removing from the slot list, decrementing the account ref count), is handled in
-                // clean_accounts() -> purge_older_root_entries()
+        purge_stats
+            .remove_cache_elapsed
+            .fetch_add(remove_cache_elapsed_across_slots, Ordering::Relaxed);
+        purge_stats
+            .num_cached_slots_removed
+            .fetch_add(num_cached_slots_removed, Ordering::Relaxed);
+        purge_stats
+            .total_removed_cached_bytes
+            .fetch_add(total_removed_cached_bytes, Ordering::Relaxed);
+    }
+
+    /// Purge the backing storage entries for the given slot, does not purge from
+    /// the cache!
+    fn purge_slot_storage_entries<'a>(
```

Comment:
1. What about calling this function `purge_dead_slots_from_storage()`, since it is called from `process_dead_slots()`?

2. I know that I'd personally find it helpful to note in the doc comment that slots in `removed_slots` must be non-rooted.

Reply:
@brooksprumo, actually, when called from the `clean_accounts-> handle_reclaims()` path, items in `removed_slots()` can be rooted if the rooted slot is dead, i.e. all of its updates have been outdated by later roots

Reply:
Oh, interesting. I think I misunderstood this part of the code:

https://github.com/solana-labs/solana/pull/17319/files/a3508c0af00c209c556b7ecf7b5120eff105bc6c#diff-1090394420d51617f3233275c2b65ed706b35b53b115fe65f82c682af8134a6fR3245-R3250

Maybe the comment can be updated?

Reply:
@brooksprumo I think I see where the confusion is :). The slot being deleted is rooted, but by the time it reaches here, the root has already been deleted from the rooted set, so that assert won't trigger. The equivalent in the old code path:

`clean_stored_dead_slots() -> finalize_dead_slot_removal() -> clean_dead_slot()` removes the roots
https://github.com/solana-labs/solana/blob/d41266e4e915cc70fdec4a3ef54d5f1289f1afaa/runtime/src/accounts_db.rs#L1889 

and then we call the equivalent `purge_storage_slots`: https://github.com/solana-labs/solana/blob/d41266e4e915cc70fdec4a3ef54d5f1289f1afaa/runtime/src/accounts_db.rs#L1893 and make the assertion check here: https://github.com/solana-labs/solana/blob/d41266e4e915cc70fdec4a3ef54d5f1289f1afaa/runtime/src/accounts_db.rs#L3198-L3200

Reply:
Updated comment to reflect this

Reply:
Ah! I get it now. Thanks for the explanation and the comments.

https://github.com/solana-labs/solana/commit/e602ebf159f0de2e72e7338ac36b4d4cf2117f03#diff-1090394420d51617f3233275c2b65ed706b35b53b115fe65f82c682af8134a6fR3190

Looks like there's a minor typo on 'alrady' in the link above.

Reply:
thanks, fixed up!